### PR TITLE
Better subset composition

### PR DESCRIPTION
**Pull request by alexnick83**
This PR attempts to address issue #575. The fix is done through `Redundant(Second)Array` instead of `Range.compose`, because it becomes impossible to infer the correct composition when reaching the innermost Memlets, where subsets are `Indices`.

### CODE REVIEW

File: dace/libraries/standard/nodes/reduce.py [hunk 1]
```diff
@@ -1,870 +1,870 @@
...
-
-    def __label__(self, sdfg, state):
-        return str(self).replace(' Axes', '\nAxes')
-
-    def validate(self, sdfg, state):
-        if len(state.in_edges(self)) != 1:
-            raise ValueError('Reduce node must have one input')
-        if len(state.out_edges(self)) != 1:
-            raise ValueError('Reduce node must have one output')
+# Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
```
Comment (alexnick83) [inline dace/libraries/standard/nodes/reduce.py:1]: @tbennun There is no change in this file (reduce.py). It is marked as changed due to an issue with git and VS code on my WSL installation.

File: dace/transformation/dataflow/redundant_array.py [hunk 2]
```diff
@@ -200,11 +250,36 @@ def apply(self, sdfg):
...
+                bset, popped = pop_dims(b_subset, b_dims_to_pop)
+
+        reduction = False
+        for e in graph.in_edges(in_array):
+            if isinstance(e.src, Reduce):
+                reduction = True
+
         # If the memlet does not cover the removed array, create a view.
-        if any(m != a for m, a in zip(a1_subset.size(), in_desc.shape)):
+        if reduction or any(m != a
```
Comment (alexnick83) [inline dace/transformation/dataflow/redundant_array.py:273]: @tbennun Due to the improvements in RedundantArray, we can now automatically remove the transient between a Reduce node and its output when `keepdims=True`. However, the Reduce node does not handle this case well (the produced code is wrong). I found how to fix it in the pure expansion, but I guess that the same problem exists in all other implementations. Therefore, I opted to put a special case here; just change the transient to a view.

File: dace/transformation/dataflow/redundant_array.py [hunk 3]
```diff
@@ -200,11 +250,36 @@ def apply(self, sdfg):
...
+                            for m, a in zip(a1_subset.size(), in_desc.shape)):
+            view_strides = in_desc.strides
+            if (b_dims_to_pop and len(b_dims_to_pop) == len(out_desc.shape) -
+                    len(in_desc.shape)):
+                view_strides = [s for i, s in enumerate(out_desc.strides)
```
Comment (alexnick83) [inline dace/transformation/dataflow/redundant_array.py:278]: @tbennun If you want a sample where `view_strides` should not be `in_desc.strides`, look at conv2d (`[N, C_out] -> [N, i, j, C_out]`).

File: dace/transformation/dataflow/redundant_array.py [hunk 4]
```diff
@@ -7,13 +7,13 @@
...
 import typing
 
-
 from dace import data, registry, subsets, dtypes, memlet as mm
 from dace.sdfg import nodes, SDFGState
 from dace.sdfg import utils as sdutil
 from dace.sdfg import graph
 from dace.transformation import transformation as pm, helpers
 from dace.config import Config
+from dace.libraries.standard import Reduce
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py:16]: Please move this into the function, it's a very slow import

**Review by tbennun — CHANGES_REQUESTED** [review]
The redundant array transformation is becoming a monster, but if all the tests pass then ok.

**Review by tbennun — APPROVED** [review]

**Pull request merged**